linode/nb/framework_resource.go

```diff
@@ -0,0 +1,269 @@
+				fmt.Sprintf("Failed to update Nodebalancer %v", plan.ID.ValueInt64()),
+				err.Error(),
+			)
+			return
+		}
+
+		resp.Diagnostics.Append(plan.parseComputedAttrs(ctx, nodebalancer)...)
+		resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	}
+}
+
+func (r *Resource) Delete(
+	ctx context.Context,
+	req resource.DeleteRequest,
+	resp *resource.DeleteResponse,
+) {
+	var data NodebalancerModel
+	client := r.Meta.Client
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	err := client.DeleteNodeBalancer(ctx, int(data.ID.ValueInt64()))
+	if err != nil {
+		if lerr, ok := err.(*linodego.Error); ok && lerr.Code == 404 {
+			resp.Diagnostics.AddWarning(
+				"Nodebalancer does not exist.",
+				fmt.Sprintf("Nodebalancer %v does not exist, removing from state.", data.ID.ValueInt64()),
+			)
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError(
+			"Failed to delete Nodebalancer",
+			err.Error(),
+		)
+		return
+	}
+}
+
+func (r *Resource) UpgradeState(context.Context) map[int64]resource.StateUpgrader {
+	return map[int64]resource.StateUpgrader{
+		0: {
+			PriorSchema:   &resourceNodebalancerV0,
+			StateUpgrader: upgradeNodebalancerResourceStateV0toV1,
+		},
+	}
+}
+
+func upgradeNodebalancerResourceStateV0toV1(
+	ctx context.Context,
+	req resource.UpgradeStateRequest,
+	resp *resource.UpgradeStateResponse,
+) {
+	var nbDataV0 nbModelV0
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &nbDataV0)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	nbDataV1 := NodebalancerModel{
+		ID:                 nbDataV0.ID,
+		Label:              nbDataV0.Label,
+		Region:             nbDataV0.Region,
+		ClientConnThrottle: nbDataV0.ClientConnThrottle,
+		Hostname:           nbDataV0.Hostname,
+		Ipv4:               nbDataV0.Ipv4,
+		Ipv6:               nbDataV0.Ipv6,
+		Created:            customtypes.RFC3339TimeStringValue{StringValue: nbDataV0.Created},
+		Updated:            customtypes.RFC3339TimeStringValue{StringValue: nbDataV0.Updated},
+		Tags:               nbDataV0.Tags,
+	}
+
+	var transferMap map[string]string
+	resp.Diagnostics.Append(nbDataV0.Transfer.ElementsAs(ctx, &transferMap, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	result := make(map[string]attr.Value)
+	in, diag := UpgradeResourceStateValue(transferMap["in"])
+	if diag != nil {
+		resp.Diagnostics.Append(diag)
+		return
+	}
+	result["in"] = in
+
+	out, diag := UpgradeResourceStateValue(transferMap["out"])
+	if diag != nil {
+		resp.Diagnostics.Append(diag)
+		return
+	}
+	result["out"] = out
+
+	total, diag := UpgradeResourceStateValue(transferMap["total"])
+	if diag != nil {
+		resp.Diagnostics.Append(diag)
+		return
+	}
+	result["total"] = total
+
+	transferObj, diags := types.ObjectValue(TransferObjectType.AttrTypes, result)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	resultList, diags := basetypes.NewListValue(
```

Comment:
I think it would be nicer if we switch all `basetypes.NewSomeValue(...)` to be `types.SomeValue(...)`, which is a higher level API.

Reply:
Good point! Updated it.